### PR TITLE
perf: cache Pipfile parse, parallelize hash/candidate lookups, harden benchmark runner

### PR DESCRIPTION
**Pull request by matteius**
## Summary

- Cut `pipenv lock` wall-clock by ~33% and user CPU by ~30% on an 86-package Pipfile (Sentry `requirements-base.txt`), measured apples-to-apples against `main` on the same machine.
- No changes to `pipenv/patched/pip/` or `pipenv/vendor/` — all wins are in editable pipenv code.
- Benchmark runner extended with repeatable ops, profiling hooks, and structured JSON output.

## Why

Profiling `pipenv lock` on a representative real-world requirement set showed three clear waste patterns in editable pipenv code:

1. The constraint file was prepared, parsed, and its PackageFinder re-built on every property access (many times per resolve).
2. `parsed_pipfile` re-read and re-parsed the same TOML document **548 times** per resolve — ~17% of wall clock spent in tomlkit.
3. Two resolver loops issued one sequential network request per resolved dep (`collect_hashes` × 149 and `find_best_candidate` × 149 inside `resolve_constraints`).

Everything below is either caching read-only state, memoising idempotent work, or parallelising independent network-bound calls.

### a96d63e7 — Cache resolver constraint work and harden benchmark runner

- Memoise `prepare_constraint_file`, `default_constraint_file`, `parsed_constraints`, `parsed_default_constraints` on `Resolver`.
- Add a `hash_finder` property so `collect_hashes` doesn't build a fresh `PackageFinder(ignore_compatibility=True)` per candidate.
- Reuse one `finder()` across the per-result loop in `resolve_constraints`.
- Drop the eager `project.environment.reverse_dependencies()` load in `process_resolver_results` — `Entry` never consumed it; update-time routines build their own.
- Benchmark runner: `TimingRecord` dataclass with rusage deltas, `--repeat` (median aggregation), `--profile` (cProfile output under `timings/`, resolver kept in-process), `--output-json`, `--no-json`, `--force-setup`.
- README / `Makefile` / `.gitignore` / CI workflow updated; CI now uploads `benchmark-results.json` alongside `stats.csv`.

### 321dde57 — Cache parsed Pipfile document to avoid repeated TOML re-parses

- `Project.parsed_pipfile` now memoises the parsed `TOMLDocument` and invalidates on mtime change.
- `Project.write_toml` explicitly drops the cache after writing the Pipfile (belt-and-suspenders for writes that share an mtime tick).
- Confirmed every Pipfile mutation path in editable pipenv code routes through `write_toml` (`add_package_to_pipfile`, `add_packages_to_pipfile_batch`, `recase_pipfile`, `ensure_proper_casing` caller, `create_pipfile`, and other helpers in `project.py`).
- `_parse_pipfile` call count in a single lock: **548 → 2**.

### d471c54d — Regression tests for Pipfile cache invalidation edge cases

- Rapid back-to-back writes on the same mtime tick are still visible on subsequent reads.
- In-place mutation of the cached document followed by `write_toml` (the `ensure_proper_casing` / `add_package_to_pipfile` pattern) persists correctly.
- `write_toml` to an unrelated path (intermediate buffer) does **not** drop the Pipfile cache.

### f056ac96 — Parallelize network-bound resolver work + small caches

- `Resolver.resolve_hashes` dispatches `collect_hashes` per-ireq on a `ThreadPoolExecutor` (max_workers ≤ 16).
- `Resolver.resolve_constraints` dispatches `find_best_candidate` per-result on a `ThreadPoolExecutor` (max_workers ≤ 8).
- Thread-safety rationale: pip's `PackageFinder._all_candidates` dict is keyed on distinct project names (atomic on different keys under the GIL); `requests.Session` is documented thread-safe; `hash_cache` delegates to pip's atomic file cache.
- Cache `Resolver.prepare_index_lookup()` — sources / index_lookup are stable after `Resolver.create()` returns.
- Hoist `canonicalize_name()` in `get_constraints_from_deps` so each dep is canonicalised once instead of twice.

## Measured impact

All numbers from `benchmarks/benchmark.py lock-warm` against an 86-package Pipfile built from Sentry's `requirements-base.txt`, same machine, same network:

| Stage | Wall | User CPU |
|---|---|---|
| `main` baseline | 22.4s | 17.7s |
| + 321dde57 (parsed_pipfile cache) | 18.8s | 14.2s |
| + f056ac96 (parallel hash/candidate + small caches) | **15.1s** | **12.3s** |
| **Branch total** | **−33%** | **−30%** |

`_parse_pipfile` invocations in the post-cache profile: 2 (down from 548). Per-pipenv-side `find_best_candidate` and `collect_hashes` calls overlap during network I/O instead of serialising.

## Test plan

- [x] `pytest tests/unit/` — 435 passed, 9 skipped (Windows), 3 pre-existing `TestTargetMarkerEnvironment` failures unrelated to this branch.
- [x] `ruff check` clean on all touched files.
- [x] `pre-commit` hooks clean on every commit.
- [x] New unit tests cover parsed_pipfile caching + invalidation, resolver constraint caching, single finder reuse in `resolve_constraints`, concurrent execution of `collect_hashes` and `find_best_candidate`, `prepare_index_lookup` caching, and the no-reverse-deps path.
- [x] Lockfile output byte-compared to pre-optimisation lock (same `_meta.hash.sha256`, same 151 resolved packages) after running on the benchmark Pipfile.
- [ ] Reviewer: please confirm behaviour on your own project's Pipfile, particularly if you use private indexes or unusual source configurations.

## Known non-issues

- At peak concurrency pip emits `Connection pool is full, discarding connection: pypi.org. Connection pool size: 10`. Cosmetic — urllib3 recycles connections transparently, no correctness impact.
- The pre-existing `TestTargetMarkerEnvironment` failures exist on `main` too (`pipenv.routines.install` no longer exposes `_python_version_for_path`). Not in scope here.

## Out of scope / future work

- The largest remaining chunk of `pipenv lock` wall-clock (~15s) is inside `pipenv/patched/pip/_internal/` — specifically `Link.from_json`, `package_finder.evaluate_link`, `Wheel.__init__`, and `parse_wheel_filename` for the ~205k candidate links enumerated across 86 packages. Memoising `Link` / `Wheel` construction by URL and short-circuiting `evaluate_link` on repeated wheel-tag-incompatible filenames would be real wins, but belong upstream in pip rather than in this branch.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/unit/test_resolver_regressions.py [hunk 1]
```diff
@@ -127,3 +131,225 @@ def test_collect_hashes_warns_once_when_all_strategies_fail(monkeypatch):
...
+    lock = threading.Lock()
+
+    def slow_find(name, specifier):
+        nonlocal concurrent, peak
+        with lock:
+            concurrent += 1
+            peak = max(peak, concurrent)
+        threading.Event().wait(timeout=0.05)
+        with lock:
+            concurrent -= 1
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_resolver_regressions.py]: @copilot apply changes based on [this feedback](https://github.com/pypa/pipenv/pull/6649#discussion_r3129429495)

File: pipenv/project.py [hunk 2]
```diff
@@ -1102,6 +1124,9 @@ def write_toml(self, data, path=None):
         formatted_data = cleanup_toml(formatted_data)
         with open(path, "w", newline=newlines) as f:
             f.write(formatted_data)
+        if Path(path).absolute() == Path(self.pipfile_location).absolute():
+            self._parsed_pipfile_cache = None
+            self._parsed_pipfile_mtime_ns = None
```
[the remark this answers is not in the document]
Comment (matteius) [reply pipenv/project.py]: @copilot apply changes based on [this feedback](https://github.com/pypa/pipenv/pull/6649#discussion_r3129529033)
